pallets/stateful-storage/src/types.rs

```diff
@@ -0,0 +1,299 @@
+			hasher.finish().to_be_bytes()[..4].try_into().expect("incorrect hash size");
+		PageHash::from_be_bytes(value_bytes)
+	}
+}
+
+/// PartialEq and Hash should be both derived or implemented manually based on clippy rules
+impl<PageDataSize: Get<u32>> Hash for Page<PageDataSize> {
+	fn hash<H: Hasher>(&self, state: &mut H) {
+		state.write(&self.data[..]);
+	}
+}
+
+/// PartialEq and Hash should be both derived or implemented manually based on clippy rules
+impl<PageDataSize: Get<u32>> PartialEq for Page<PageDataSize> {
+	fn eq(&self, other: &Self) -> bool {
+		self.data.eq(&other.data)
+	}
+}
+
+impl<PageDataSize: Get<u32>> From<BoundedVec<u8, PageDataSize>> for Page<PageDataSize> {
+	fn from(bounded: BoundedVec<u8, PageDataSize>) -> Self {
+		Self { data: bounded }
+	}
+}
+
+impl<PageDataSize: Get<u32>> TryFrom<Vec<u8>> for Page<PageDataSize> {
+	type Error = ();
+
+	fn try_from(data: Vec<u8>) -> Result<Self, Self::Error> {
+		let bounded: BoundedVec<u8, PageDataSize> = BoundedVec::try_from(data).map_err(|_| ())?;
+		Ok(Page::from(bounded))
+	}
+}
+
+impl<PageDataSize: Get<u32>> Page<PageDataSize> {
+	/// Applies all actions to specified page and returns the updated page
+	/// This has O(n) complexity when n is the number of all the bytes in that itemized storage
+	pub fn apply_item_actions<T: Config>(
+		&self,
+		actions: &[ItemAction<T::MaxItemizedBlobSizeBytes>],
+	) -> Result<Self, PageError> {
+		let mut parsed = self.parse_as_itemized(true)?;
+
+		let mut updated_page_buffer = Vec::with_capacity(parsed.page_size);
+		let mut add_buffer = Vec::new();
+
+		for action in actions {
+			match action {
+				ItemAction::Delete { index } => {
+					ensure!(
+						parsed.items.contains_key(&index),
+						PageError::InvalidAction("item index is invalid")
+					);
+					parsed.items.remove(&index);
+				},
+				ItemAction::Add { data } => {
+					let header = ItemHeader {
+						payload_len: data
+							.len()
+							.try_into()
+							.map_err(|_| PageError::InvalidAction("invalid payload size"))?,
+					};
+					add_buffer.extend_from_slice(&header.encode()[..]);
+					add_buffer.extend_from_slice(&data[..]);
+				},
+			}
+		}
+
+		// since BTreeMap is sorted by key, all items will be kept in their existing order
+		for (_, slice) in parsed.items.iter() {
+			updated_page_buffer.extend_from_slice(slice);
+		}
+		updated_page_buffer.append(&mut add_buffer);
+
+		Page::<PageDataSize>::try_from(updated_page_buffer).map_err(|_| PageError::PageSizeOverflow)
+	}
+
+	/// Parses all the items inside an ItemPage
+	/// This has O(n) complexity when n is the number of all the bytes in that itemized storage
+	pub fn parse_as_itemized(&self, include_header: bool) -> Result<ParsedItemPage, PageError> {
+		let mut count = 0u16;
+		let mut items = BTreeMap::new();
+		let mut offset = 0;
+		while offset < self.data.len() {
+			ensure!(
+				offset + ItemHeader::max_encoded_len() <= self.data.len(),
+				PageError::ErrorParsing("wrong header size")
```

Comment:
maybe we can remove these strings  and create another enum like
```
pub enum <Something>Error {
   InvalidHeaderSize,
   InvalidPayloadSize,
   HeaderDecodeFail,
} 
```

Reply:
Thought about that but then I'll have to create another enum for actions also and include both in PageError enum. Not sure if it worths the changes. I also print these in case of failure since it will allow us to know what is a wrong with storage later.